File: advent of code/2022/day5.ts
import { readInput } from "./utils.ts"

export function day5(test = false) {
  const input = readInput(5, test)
  const [rawInitialStacks, rawProcedures] = input.split("\n\n").map((part) => part.split("\n"))

  const initialStacks = rawInitialStacks
    .reverse()
    .slice(1) // skip footer row
    .map((row) => {
      return row
        .replace(/[\[\]]*/g, "") // remove square brackets
        .replace(/\W{4}/g, " ") // collapse 4-spaces into a single space
        .split(" ")
    })

  const procedures: any = rawProcedures.map((row) => {
    return Array.from(row.matchAll(/\d+/g)).flat()
  })

  const stacks1: string[][] = []
  for (const row of initialStacks) {
    row.forEach((col, i) => {
      stacks1[i + 1] ??= []
      if (col !== "") {
        stacks1[i + 1].push(col)
      }
    })
  }
  const stacks2: string[][] = structuredClone(stacks1)

  for (const [num, src, dst] of procedures) {
    for (let i = 0; i < num; i++) {
      stacks1[dst].push(stacks1[src].pop()!)
    }
    const removed = stacks2[src].splice(stacks2[src].length - num, num)
    stacks2[dst].push(...removed)
  }

  return [stacks1.map((a) => a.at(-1)).join(""), stacks2.map((a) => a.at(-1)).join("")]
}

File: advent of code/2022/ocaml/lib/day5.ml
open Core

let solve ~(file : string) =
  let input = Utils.read_file file in
  let raw_initial_stacks, raw_procedures =
    Utils.split input "\n\n" |> Utils.tuple_exn
  in
  let stacks =
    raw_initial_stacks |> String.split_lines |> List.drop_last_exn
    |> List.map ~f:String.to_array
    |> List.to_array |> Array.transpose_exn
    |> Array.filteri ~f:(fun i _ -> i % 4 = 1)
    |> Array.map ~f:(Array.filter ~f:(fun c -> not (Char.is_whitespace c)))
    |> Array.map ~f:Array.to_list
  in
  let procedures =
    raw_procedures |> String.split_lines
    |> List.map ~f:Utils.extract_numbers
    |> List.map ~f:(function
         | [ num; src; dst ] -> (num, src, dst)
         | _ -> failwith (sprintf "Unexpected input"))
  in

  let stacks1, stacks2 = (Array.copy stacks, Array.copy stacks) in
  List.iter procedures ~f:(fun (num, src, dst) ->
      let src, dst = (src - 1, dst - 1) in
      let moved, new_src = List.split_n stacks1.(src) num in
      stacks1.(src) <- new_src;
      stacks1.(dst) <- List.rev moved @ stacks1.(dst);

      let moved, new_src = List.split_n stacks2.(src) num in
      stacks2.(src) <- new_src;
      stacks2.(dst) <- moved @ stacks2.(dst));

  let all_last stack =
    Array.map stack ~f:List.hd_exn |> Array.to_list |> String.of_char_list
  in
  let part1 = all_last stacks1 in
  let part2 = all_last stacks2 in
  (part1, part2)

let%test_unit _ =
  [%test_eq: string * string] (solve ~file:"day5-example.txt") ("CMZ", "MCD")
